File: app/src/main/java/com/hover/stax/settings/BiometricChecker.java
package com.hover.stax.settings;

import android.app.KeyguardManager;
import android.content.Context;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.biometric.BiometricPrompt;

import com.hover.sdk.actions.HoverAction;
import com.hover.stax.R;
import com.hover.stax.utils.Utils;

import java.util.concurrent.Executor;
import java.util.concurrent.Executors;

import static androidx.biometric.BiometricManager.Authenticators.BIOMETRIC_STRONG;
import static androidx.biometric.BiometricManager.Authenticators.BIOMETRIC_WEAK;
import static androidx.biometric.BiometricManager.Authenticators.DEVICE_CREDENTIAL;


public class BiometricChecker extends BiometricPrompt.AuthenticationCallback {
    private AppCompatActivity a;
    private AuthListener listener;
    private HoverAction action;

    public BiometricChecker(AuthListener authListener, AppCompatActivity activity) {
        listener = authListener;
        a = activity;
    }

    public void startAuthentication(HoverAction act) {
        action = act;
        if (!((KeyguardManager) a.getSystemService(Context.KEYGUARD_SERVICE)).isKeyguardSecure()) {
            listener.onAuthSuccess(action);
            return;
        }

        Executor newExecutor = Executors.newSingleThreadExecutor();
        final BiometricPrompt myBiometricPrompt = new BiometricPrompt(a, newExecutor, this);

        final BiometricPrompt.PromptInfo promptInfo =
                new BiometricPrompt.PromptInfo.Builder()
                        .setTitle(a.getString(R.string.auth_title))
                        .setAllowedAuthenticators(BIOMETRIC_STRONG | BIOMETRIC_WEAK | DEVICE_CREDENTIAL)
                        .build();

        myBiometricPrompt.authenticate(promptInfo);
    }

    @Override
    public void onAuthenticationError(int errorCode, @NonNull CharSequence error) {
        super.onAuthenticationError(errorCode, error);
        if (errorCode == BiometricPrompt.ERROR_NO_BIOMETRICS) {
            listener.onAuthError(error.toString());
        } else Utils.logAnalyticsEvent(a.getString(R.string.biometrics_not_setup), a.getBaseContext());
    }

    @Override
    public void onAuthenticationSucceeded(@NonNull BiometricPrompt.AuthenticationResult result) {
        super.onAuthenticationSucceeded(result);
        Utils.logAnalyticsEvent(a.getString(R.string.biometrics_succeeded), a.getBaseContext());
        listener.onAuthSuccess(action);
    }

    @Override
    public void onAuthenticationFailed() {
        super.onAuthenticationFailed();
        Utils.logAnalyticsEvent(a.getString(R.string.biometrics_failed), a.getBaseContext());
    }

    public interface AuthListener {
        void onAuthError(String error);

        void onAuthSuccess(HoverAction action);
    }
}

File: app/src/main/java/com/hover/stax/settings/BiometricChecker.kt
/*
 * Copyright 2022 Stax
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.hover.stax.settings

import android.app.KeyguardManager
import android.content.Context
import androidx.appcompat.app.AppCompatActivity
import androidx.biometric.BiometricManager
import androidx.biometric.BiometricManager.Authenticators.BIOMETRIC_STRONG
import androidx.biometric.BiometricManager.Authenticators.BIOMETRIC_WEAK
import androidx.biometric.BiometricManager.Authenticators.DEVICE_CREDENTIAL
import androidx.biometric.BiometricPrompt
import androidx.core.content.ContextCompat
import com.hover.sdk.actions.HoverAction
import com.hover.stax.R
import com.hover.stax.core.AnalyticsUtil

class BiometricChecker(private val authListener: AuthListener, val activity: AppCompatActivity) :
    BiometricPrompt.AuthenticationCallback() {

    private var action: HoverAction? = null

    fun startAuthentication(a: HoverAction?) {
        action = a

        if (!(activity.getSystemService(Context.KEYGUARD_SERVICE) as KeyguardManager).isKeyguardSecure) {
            authListener.onAuthSuccess(action)
            return
        }

        val biometricsManager = BiometricManager.from(activity)
        when (biometricsManager.canAuthenticate(BIOMETRIC_STRONG or BIOMETRIC_WEAK or DEVICE_CREDENTIAL)) {
            BiometricManager.BIOMETRIC_SUCCESS -> launchBiometricsPrompt()
            BiometricManager.BIOMETRIC_ERROR_NO_HARDWARE,
            BiometricManager.BIOMETRIC_ERROR_HW_UNAVAILABLE,
            BiometricManager.BIOMETRIC_ERROR_NONE_ENROLLED -> authListener.onAuthSuccess(action)
            else -> authListener.onAuthSuccess(action)
        }
    }

    private fun launchBiometricsPrompt() {
        val executor = ContextCompat.getMainExecutor(activity)
        val biometricPrompt = BiometricPrompt(activity, executor, this)
        val promptInfo = BiometricPrompt.PromptInfo.Builder()
            .setTitle(activity.getString(R.string.auth_title))
            .setAllowedAuthenticators(BIOMETRIC_STRONG or BIOMETRIC_WEAK or DEVICE_CREDENTIAL)
            .build()
        biometricPrompt.authenticate(promptInfo)
    }

    override fun onAuthenticationSucceeded(result: BiometricPrompt.AuthenticationResult) {
        super.onAuthenticationSucceeded(result)
        com.hover.stax.core.AnalyticsUtil.logAnalyticsEvent(
            activity.getString(R.string.biometrics_succeeded),
            activity.baseContext
        )
        authListener.onAuthSuccess(action)
    }

    override fun onAuthenticationError(errorCode: Int, errString: CharSequence) {
        super.onAuthenticationError(errorCode, errString)
        if (errorCode == BiometricPrompt.ERROR_NO_BIOMETRICS)
            authListener.onAuthError(errString.toString())
        else
            com.hover.stax.core.AnalyticsUtil.logAnalyticsEvent(
                activity.getString(R.string.biometrics_not_setup),
                activity.baseContext
            )
    }

    override fun onAuthenticationFailed() {
        super.onAuthenticationFailed()
        com.hover.stax.core.AnalyticsUtil.logAnalyticsEvent(
            activity.getString(R.string.biometrics_failed),
            activity.baseContext
        )
    }

    interface AuthListener {
        fun onAuthError(error: String)
        fun onAuthSuccess(action: HoverAction?)
    }
}